core/modules/replica/ChunkLocker.h

```diff
@@ -0,0 +1,222 @@
+     */
+    ChunksByOwners locked(std::string const& owner=std::string()) const;
+
+    /**
+     * Lock a chunk to a specific owner and return 'true' of the operation
+     * was successfull.
+     *
+     * NOTE: if an attempt to claim the chunk by the same owner which already
+     *       owns it is made then the method will return 'true'.
+     *
+     * @param chunk - a chunk to be locked
+     * @param owner - an identifier of an owner claiming thr chunk
+     *
+     * @throw std::invalid_argument - if the ownerId is an empty string
+     */
+    bool lock(Chunk const&       chunk,
+              std::string const& owner);
+
+    /**
+     * Release a chunk and return 'true' if the operation was successfull
+     *
+     * @param chunk - a chunk to be released
+     */
+    bool release(Chunk const& chunk);
+
+    /**
+     * Release a chunk and return 'true' if the operation was successfull.
+     * In the latter case also set an identifier of an owner which previously
+     * 'claimed' the chunk.
+     *
+     * @param chunk - a chunk to be released
+     * @param owner - a reference to a string which will be initialized with
+     *                an identifier of an owner which had a claim on the chunk
+     *                at a time of the method call if the chunk was found locked
+     */
+    bool release(Chunk const& chunk,
+                 std::string& owner);
+
+    /**
+     * Release all chunks which were found claimed by the specified owner
+     * and return a collection of those chunks
+     *
+     * @throw std::invalid_argument - if the owner is an empty string
+     */
+    std::vector<Chunk> release(std::string const& owner);
+
+private:
+
+    /**
+     * Release a chunk and return 'true' if the operation was successfull.
+     * In the latter case also set an identifier of an owner which previously
+     * 'claimed' the chunk.
+     *
+     * NOTE: this metod is not thread-safe. It's up to its callers to ensure
+     *       proper synchronization context before invoking the method.
+     *
+     * @param mLock - a lock on a mutex must be made before calling this method
+     * @param chunk - a chunk to be released
+     * @param owner - a reference to a string which will be initialized with
+     *                an identifier of an owner which had a claim on the chunk
+     *                at a time of the method call if the chunk was found locked
+     */
+    bool releaseImpl(util::Lock const& mLock,
+                     Chunk const& chunk,
+                     std::string& owner);
+
+private:
+
+    /// Mapping a chunk to its "owner" (the one which holds the lock)
+    std::map<Chunk, std::string> _chunk2owner;
+
+    /// Mapping an owner to a list of chunks "clamed" by that owner
+    /// NOTE: using the list container for the better performance of
+    ///       the insert/erase operations over the lists of chunks
+    ChunksByOwners _owner2chunks;
```

Comment:
you might consider using boost::multiindex instead of keeping 2 lists. Verifying synchronization of these lists in the cc file is giving me a headache 😉

Reply:
That's an interesting idea! I must admit, I've never heard about boost::multi index. I will investigate the option.

Reply:
I have browse through the documentation on boost::multi index, and I really didn't like it! It's so complicated and obscure. I would rather simplify the implementation of the ChunkLocker class by getting rid of the secondary index '_owner2chunks'. And indeed, after re-examining the code, it appears to me this removal won't lead to any noticeable sacrifices in terms of performance ro clarity of the code.